graphql_compiler/compiler/ir_lowering_match/between_lowering.py

```diff
@@ -37,54 +42,68 @@ def _extract_conjuction_elements_from_expression(expression):
+def _construct_field_operator_expression_dict(
+    expression_list: List[Expression],
+) -> Tuple[
+    Dict[str, Dict[str, List[BinaryComposition]]], Dict[str, GraphQLOutputType], Deque[Expression],
+]:
     """Construct a mapping from local fields to specified operators, and corresponding expressions.
 
     Args:
         expression_list: list of expressions to analyze
 
     Returns:
-        tuple (field_name_to_expressions, field_name_to_type, remaining_expression_list), containing
+        tuple (field_name_to_expressions, field_name_to_type, remaining_expression_deque) containing
         - field_name_to_expressions:
             dict mapping local field names to "operator -> list of BinaryComposition" dictionaries,
             for each BinaryComposition operator involving the LocalField
         - field_name_to_type:
             dict mapping local field names to field type
-        - remaining_expression_list:
-            list of remaining expressions that were *not*
+        - remaining_expression_deque:
+            deque of remaining expressions that were *not*
             BinaryCompositions on a LocalField using any of the between operators
     """
     between_operators = ("<=", ">=")
     inverse_operator = {">=": "<=", "<=": ">="}
-    field_name_to_expressions = {}
-    field_name_to_type = {}
-    remaining_expression_list = deque([])
+    field_name_to_expressions: Dict[str, Dict[str, List[BinaryComposition]]] = {}
+    field_name_to_type: Dict[str, GraphQLOutputType] = {}
+    remaining_expression_deque: Deque[Expression] = deque([])
 
     for expression in expression_list:
-        if all(
-            (
-                isinstance(expression, BinaryComposition),
-                expression.operator in between_operators,
-                isinstance(expression.left, LocalField) or isinstance(expression.right, LocalField),
+        if (
+            isinstance(expression, BinaryComposition)
+            and expression.operator in between_operators
+            and (
+                isinstance(expression.left, LocalField) or isinstance(expression.right, LocalField)
             )
         ):
+            # Ensure we rewrite the BinaryComposition in such a way that the left expression
+            # is always a LocalField.
             if isinstance(expression.right, LocalField):
                 new_operator = inverse_operator[expression.operator]
                 new_expression = BinaryComposition(new_operator, expression.right, expression.left)
             else:
                 new_expression = expression
-            field_name = new_expression.left.field_name
-            field_type = new_expression.left.field_type
+
+            left_clause = new_expression.left
+            if not isinstance(left_clause, LocalField):
+                raise AssertionError(
+                    f"Expected left_clause to be a LocalField, but instead found {left_clause}. "
```

Comment:
tiny nit: I wouldn't mention the "left_clause" variable name in the error message since we could easily rename the left_clause variable name and forget to update the error message here.

Reply:
I'm not sure what a way to refer to this value that doesn't have "we could rename the variable" risk. The same is true of `new_expression.left` and of the value that `new_expression` comes from. Do you have a preference for a particular resolution?

Reply:
Maybe "Expected a LocalField, but got {left_clause"} instead. This is a bug"?

Reply:
To be honest, I'm not a huge fan — yes, that protects from the renaming risk, but at the cost of what feels to me like worse readability today. I think I'd prefer the variable to be named, raise a more informative error, and watch in code reviews for missing renamings, than be protected from renaming risk at the cost of less-informative errors.